Core/Src/system.c

```diff
@@ -79,22 +96,53 @@ void system_main(System *sys) {
 
 void commandHandler(System *sys) {
 	uint8_t command;
-	uint8_t status_error = pop_ring_buffer(sys->recieve_ring_buffer, &command);
+	uint8_t status_error = pop_ring_buffer(sys->receive_ring_buffer, &command);
 	if (!status_error) {
-		switch (command) {
-		case LED1_5Hz:
-			sys->LED1_timer = sys->h_timer1;
-			break;
-		case LED1_1Hz:
-			sys->LED1_timer = sys->h_timer2;
-			break;
-		case LED2_5Hz:
-			sys->LED2_timer = sys->h_timer1;
-			break;
-		case LED2_1Hz:
-			sys->LED2_timer = sys->h_timer2;
-			break;
+
+		if (command == sys->LED1_State) {
+			sys->LED1_State = OFF;
+			HAL_GPIO_WritePin(sys->h_GPIO_LED, sys->LEDPin1, GPIO_PIN_RESET);
+			HAL_UART_Transmit_IT(&huart3, (uint8_t*)"Turning LED1 OFF.\n", sizeof("Turning LED1 OFF.\n"));
+
+		}
+		else if (command == sys->LED2_State) {
+			sys->LED2_State = OFF;
+			HAL_GPIO_WritePin(sys->h_GPIO_LED, sys->LEDPin2, GPIO_PIN_RESET);
+			HAL_UART_Transmit_IT(&huart3, (uint8_t*)"Turning LED2 OFF.\n", sizeof("Turning LED2 OFF.\n"));
+
+		}
+		else {
+			switch (command) {
+			case LED1_5Hz:
+				sys->LED1_timer = sys->h_timer1; // use h_timer1 as the LED1 timer
+				sys->LED1_State = command;
+				HAL_UART_Transmit_IT(&huart3, (uint8_t*)"Blinking LED1 at 5 Hz.\n", sizeof("Blinking LED1 at 5 Hz.\n"));
+				break;
+			case LED1_1Hz:
+				sys->LED1_timer = sys->h_timer2; // use h_timer2 as the LED1 timer
+				sys->LED1_State = command;
+				HAL_UART_Transmit_IT(&huart3, (uint8_t*)"Blinking LED1 at 1 Hz.\n", sizeof("Blinking LED1 at 1 Hz.\n"));
+				break;
+			case LED2_5Hz:
+				sys->LED2_timer = sys->h_timer1; // use h_timer1 as the LED2 timer
+				sys->LED2_State = command;
+				HAL_UART_Transmit_IT(&huart3, (uint8_t*)"Blinking LED2 at 5 Hz.\n", sizeof("Blinking LED2 at 5 Hz.\n"));
+				break;
+			case LED2_1Hz:
+				sys->LED2_timer = sys->h_timer2; // use h_timer2 as the LED2 timer
+				sys->LED2_State = command;
+				HAL_UART_Transmit_IT(&huart3, (uint8_t*)"Blinking LED2 at 1 Hz.\n", sizeof("Blinking LED2 at 1 Hz.\n"));
+				break;
+
+			case HANDSHAKE_ARE_YOU_THERE:
+				HAL_UART_Transmit_IT(&huart3, (uint8_t*)"Are you still there?\n", sizeof("Are you still there?\n"));
+				HAL_GPIO_WritePin(sys->h_GPIO_LED, sys->LEDPin1, GPIO_PIN_RESET);
+				HAL_GPIO_WritePin(sys->h_GPIO_LED, sys->LEDPin2, GPIO_PIN_RESET);
+				break;
+			}
 		}
+		HAL_UART_Receive_IT(&huart3, sys->incomingByte, SIZE_INCOMING_DATA); // allow for more receives
```

Comment:
I might even say it would be valid to put the next function call in the callback function too, if we can assume commands will come in faster than human inputs.  But since we are limited to corporeal reflexes, I doubt that a user will manage to beat the processor into inputting another command before the next receive interrupt call if called within the main loop